apps/schematic/api/Dockerfile

```diff
@@ -1,41 +1,70 @@
-FROM python:3.10.12-slim-buster
+FROM tiangolo/uwsgi-nginx-flask:python3.10
 
-ENV APP_DIR=/opt/app
+# add version tag as a build argument
+ARG TAG
 
-SHELL ["/bin/bash", "-euxo", "pipefail", "-c"]
+# add label 
+LABEL maintainer="Milen Nikolov, Andrew Lamb, Mialy DeFelice, Gianna Jordan, Lingling Peng"
 
-# hadolint ignore=DL3008
-RUN apt-get update -qq -y \
-    && apt-get install --no-install-recommends -qq -y \
-        build-essential \
-        gosu \
-        libpcre3 \
-        libpcre3-dev \
-        python3-dev \
-    && apt-get -y autoclean \
-    && apt-get -y autoremove \
-    && rm -rf /var/lib/apt/lists/*
+# the environment variables defined here are the default
+# and can be overwritten by docker run -e VARIABLE = XX
+# or can be overwritten by .env when using docker compose
+ENV PYTHONFAULTHANDLER=1 \
+    PYTHONUNBUFFERED=1 \
+    PYTHONHASHSEED=random \
+    PIP_NO_CACHE_DIR=off \
+    PIP_DISABLE_PIP_VERSION_CHECK=on \
+    PIP_DEFAULT_TIMEOUT=200 \
+    POETRY_VERSION=1.3.0 \
+    APP_PARENT_DIR=/app \ 
+    NGINX_CONFIG=/etc/nginx/conf.d \
+    APP_DIR=/app/app \
+    ROOT=/ \
+    UWSGI_INI=/app/uwsgi.ini \ 
+    NGINX_WORKER_PROCESSES=1 \
+    VERSION=$TAG
 
-WORKDIR ${APP_DIR}
-COPY schematic_api schematic_api/
-COPY pyproject.toml poetry.lock uwsgi.ini ./
+# run open ssl and generate certificate
+RUN apt update && \
+    apt-get install openssl && \
+    openssl req -x509 -nodes -days 365 \
```

Comment:
- Is it OK to make the self-signed certificate publicly available (assuming a public image)?
- Is it OK for the certificate to change each time the image is rebuilt?

Reply:
Here's a breakdown of the considerations for deciding whether to add a self-signed certificate to a Docker image or mount it to the container:

Adding to the Image:

Pros:

    Self-contained: The certificate is embedded within the image, making it portable and usable across environments.
    Simplified deployments: No need to manage external volumes or mounts during container deployment.

Cons:

    Image updates: Requires rebuilding the image if the certificate needs to be updated or replaced.
    Security: Less secure if the image is shared publicly, as the certificate is exposed.

Mounting to the Container:

Pros:

    Flexibility: Allows for easy updates or replacement of the certificate without rebuilding the image.
    Security: Keeps the certificate external to the image, reducing exposure if the image is shared.
    Centralized management: Can be managed independently of the image, simplifying certificate updates across multiple containers.

Cons:

    Mounting complexity: Requires managing external volumes or mounts, adding complexity to deployment.
    Potential permissions issues: Ensure proper permissions are set for the mounted certificate within the container.

Best Practices:

    For sensitive environments: Prioritize security and flexibility by mounting the certificate.
    For frequent certificate updates: Mounting is more efficient.
    For simple deployments and self-contained images: Adding to the image might be suitable.
    Consider using Docker Secrets: Provides a secure mechanism for managing sensitive information like certificates.

Additional Considerations:

    Certificate expiration: Implement a process for monitoring and renewing certificates regardless of the chosen approach.
    Image distribution: If sharing images publicly, avoid embedding sensitive certificates.
    Volume drivers: Choose a suitable volume driver for mounting certificates, considering performance and security implications.

Ultimately, the best approach depends on your specific requirements and security considerations.

Reply:
Thanks for the feedbacks. I did more googling and noticed that a lot people mounted the certificate instead of including the certificate in the docker image. But I am not sure how they managed deployment. Let me reach out to IT and see if adding a self-signed certificate to a Docker image creates security concerns

Reply:
@tschaffter thanks for your feedback. I have updated the Dockerfile and made sure that we pass certificate and private key as environment variables.